### PR TITLE
CTOOLS-329: add timeout and rate limit retries config

### PR DESCRIPTION
**Pull request by cg-finbourne**
# Pull Request Checklist

- [ ] Read the [contributing guidelines](../blob/master/docs/CONTRIBUTING.md)
- [ ] Tests pass
- [ ] Raised the PR against the `develop` branch

# Description of the PR

Describe the code changes for the reviewers, explain the solution you have provided and how it fixes the issue

### CODE REVIEW

File: generate/templates/ApiClient.mustache [hunk 1]
```diff
@@ -206,6 +206,59 @@ namespace {{packageName}}.Client
         }
     }
 
+    /// <summary>
+    /// An interface for the required RestClient methods
+    /// </summary>
+    public interface IRestClientWrapper : IRestClient
```
Comment (cg-finbourne) [inline generate/templates/ApiClient.mustache:212]: this was added to make testing possible without actually calling an api

File: generate/templates/ApiClient.mustache [hunk 2]
```diff
@@ -262,12 +316,23 @@ namespace {{packageName}}.Client
         /// <param name="CreateHttpMessageHandler">A function to create a HttpMessageHandler for each REST request</param>
         /// <param name="disposeHandler">Should the HttpMessageHandler be disposed of after each request</param>
         /// <exception cref="ArgumentException"></exception>
-        public ApiClient(string basePath, Func<RestClientOptions, HttpMessageHandler>? CreateHttpMessageHandler = null, bool disposeHandler = true){
+        public ApiClient(string basePath, Func<RestClientOptions, HttpMessageHandler>? CreateHttpMessageHandler = null, bool disposeHandler = true, Func<RestClientOptions, IReadableConfiguration, IRestClientWrapper> createRestClientFunc = null){
```
Comment (cg-finbourne) [inline generate/templates/ApiClient.mustache:319]: similarly added for testing

File: tests/add-tests.sh [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+}
+trap 'failure ${LINENO} "$BASH_COMMAND"' ERR
+
+sdk_dir=$1
+
+# copy the tests dir to the sdk dir
+cp -r tests/Tests "$sdk_dir/sdk"
+
+# make all necessary replacements
+upper_case_application_name="$(echo "$APPLICATION_SHORT_NAME" | tr '[:lower:]' '[:upper:]')"; \
```
Comment (cg-finbourne) [inline tests/add-tests.sh:18]: I left a comment but it seems to have disappeared. To repeat - this sed'ing is something we should get rid of. At this point it would actually be less brittle and more readable to template these test files. It would have the added benefit that we could see the interpolated final state test files in src in gitlab. But we should do in a separate mr

File: generate/templates/RequestOptions.mustache [hunk 4]
```diff
@@ -0,0 +1,97 @@
+{{>partial_header}}
+
```
Comment (cg-finbourne) [inline generate/templates/RequestOptions.mustache:2]: this is the diff 
image

File: generate/templates/RetryConfiguration.mustache [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        /// Retry policy - if set will be used over any policy returned from GetRetryPolicyFunc
+        /// </summary>
+        public static Policy<HttpResponseMessage>? RetryPolicy { get; set; }
+
+        /// <summary>
+        /// Async retry policy - if set will be used over any policy returned from GetAsyncRetryPolicyFunc
+        /// </summary>
+        public static AsyncPolicy<HttpResponseMessage>? AsyncRetryPolicy { get; set; }
+{{/useHttpClient}}
+    }
```
Comment (cg-finbourne) [inline generate/templates/RetryConfiguration.mustache:61]: this is the diff 
image

File: generate/templates/other/PollyApiRetryHandler.mustache [hunk 6]
```diff
@@ -26,8 +29,6 @@ namespace {{packageName}}.Extensions
         /// </summary>
         public const int DefaultNumberOfRetries = 2;
 
-        private const int RateLimitRetryCount = 3;
```
Comment (cg-finbourne) [inline generate/templates/other/PollyApiRetryHandler.mustache:29]: this is not set by configuration

File: generate/templates/netcore_project.additions.mustache [hunk 7]
```diff
@@ -1,7 +1,8 @@
 
   <PropertyGroup>
     <WarningsAsErrors></WarningsAsErrors>
-    <NoWarn>1701;1702;CS0472;CS1573;CS0108;CS8073</NoWarn>
+    <!-- REMOVE NU1902 ONCE CTOOLS-333 complete -->
+    <NoWarn>1701;1702;CS0472;CS1573;CS0108;CS8073;NU1902</NoWarn>
```
Comment (cg-finbourne) [inline generate/templates/netcore_project.additions.mustache:5]: added this because rider keeps error'ing about this, even though `dotnet build` doesn't

File: generate/templates/RequestOptions.mustache [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        /// <summary>
+        /// If request should be authenticated with OAuth.
+        /// </summary>
+        public bool OAuth { get; set; }
+        {{/hasOAuthMethods}}
+        
+        /// <summary>
+        /// The client timeout in milliseconds. If left unset the default timeout will be used
+        /// </summary>
+        public int? TimeoutMs { get; set; }
```
Comment (cg-finbourne) [inline generate/templates/RequestOptions.mustache:75]: new field

File: tests/Tests/Unit/ApiExceptionExtensionsTest.cs [hunk 9]
```diff
@@ -7,32 +7,6 @@ namespace Finbourne.Sdk.Extensions.Tests.Unit
...
-            ApiException apiEx = new ApiException((int)HttpStatusCode.BadRequest, "message");
-            Assert.IsTrue(apiEx.TryGetValidationProblemDetails(out LusidValidationProblemDetails _));
-        }
-
-        [Test]
-        public void TryGetValidationProblemDetails_HttpStatusCode_BadGateway_Returns_False()
-        {
-            ApiException apiEx = new ApiException((int)HttpStatusCode.BadGateway, "message");
-            Assert.IsFalse(apiEx.TryGetValidationProblemDetails(out LusidValidationProblemDetails _));
-        }
```
Comment (cg-finbourne) [inline tests/Tests/Unit/ApiExceptionExtensionsTest.cs:34]: these were removed unfortunately because the IsValidationProblem and TryGetValidationProblemDetails methods are only generated if generate_validation_exception_code is true (true when LusidValidationProblemDetails exists in the openapi spec - see [here](https://github.com/finbourne/lusid-sdk-generator-csharp/blob/feature/CTOOLS-329/generate/generate.sh#L57)). So for luminesce for example these methods don't exist and there is a compile error.

This can be fixed when/if we template these test files rather than copying them

File: generate/templates/RequestOptions.mustache (hunk 8 above)
Comment (AndrewNeudegg) [reply generate/templates/RequestOptions.mustache:75]: is it nullable?

File: generate/templates/RequestOptions.mustache [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        
+        /// <summary>
+        /// The client timeout in milliseconds. If left unset the default timeout will be used
+        /// </summary>
+        public int? Timeout { get; set; }
+        
+        /// <summary>
+        /// The number of retries when being rate limited. If left unset the default number of rate limit retries will be used
+        /// </summary>
+        public int? RateLimitRetries { get; set; }
```
[the remark this answers is not in the document]
Comment (AndrewNeudegg) [reply generate/templates/RequestOptions.mustache:80]: is it nullable?

**Review by AndrewNeudegg — APPROVED** [review]
